File: apps/mobile/src/scan/tracker.ts
import type { Box, Detection, Track } from './types';
import { MAX_DETECTIONS, validDetection } from './decoder';

export function intersectionOverUnion(a: Box, b: Box): number {
  const left = Math.max(a.x, b.x),
    top = Math.max(a.y, b.y);
  const right = Math.min(a.x + a.width, b.x + b.width),
    bottom = Math.min(a.y + a.height, b.y + b.height);
  const intersection = Math.max(0, right - left) * Math.max(0, bottom - top);
  const union = a.width * a.height + b.width * b.height - intersection;
  return union <= 0 ? 0 : intersection / union;
}

export function nonMaxSuppression(
  detections: readonly Detection[],
  overlapThreshold = 0.45,
): Detection[] {
  if (!Number.isFinite(overlapThreshold) || overlapThreshold <= 0 || overlapThreshold > 1)
    throw new RangeError('overlapThreshold must be greater than zero and at most one');
  const result: Detection[] = [];
  const remaining = detections
    .slice(0, MAX_DETECTIONS)
    .filter(validDetection)
    .map((d) => ({ ...d, box: { ...d.box } }))
    .sort((a, b) => b.confidence - a.confidence);
  while (remaining.length) {
    const candidate = remaining.shift()!;
    result.push(candidate);
    for (let i = remaining.length - 1; i >= 0; i -= 1) {
      if (
        remaining[i].label === candidate.label &&
        intersectionOverUnion(candidate.box, remaining[i].box) >= overlapThreshold
      )
        remaining.splice(i, 1);
    }
  }
  return result;
}

export interface TrackerOptions {
  matchIou?: number;
  maxMissedFrames?: number;
}

export class CropTracker {
  private tracks: Track[] = [];
  private nextId = 1;
  private readonly matchIou: number;
  private readonly maxMissedFrames: number;
  constructor(options: TrackerOptions = {}) {
    this.matchIou = options.matchIou ?? 0.15;
    this.maxMissedFrames = options.maxMissedFrames ?? 4;
    if (!Number.isFinite(this.matchIou) || this.matchIou <= 0 || this.matchIou >= 1)
      throw new RangeError('matchIou must be greater than zero and less than one');
    if (
      !Number.isSafeInteger(this.maxMissedFrames) ||
      this.maxMissedFrames < 0 ||
      this.maxMissedFrames > 30
    )
      throw new RangeError('maxMissedFrames must be an integer between zero and thirty');
  }
  update(rawDetections: readonly Detection[]): Track[] {
    const detections = nonMaxSuppression(rawDetections),
      used = new Set<number>();
    const updated = this.tracks.map((track) => {
      let bestIndex = -1,
        bestIou = this.matchIou;
      detections.forEach((detection, index) => {
        if (used.has(index) || detection.label !== track.label) return;
        const iou = intersectionOverUnion(track.box, detection.box);
        if (iou > bestIou) {
          bestIou = iou;
          bestIndex = index;
        }
      });
      if (bestIndex < 0) return { ...track, missedFrames: track.missedFrames + 1 };
      used.add(bestIndex);
      return { ...detections[bestIndex], id: track.id, missedFrames: 0 };
    });
    const fresh = detections
      .filter((_, index) => !used.has(index))
      .map((detection) => ({ ...detection, id: this.nextId++, missedFrames: 0 }));
    this.tracks = [
      ...updated.filter((track) => track.missedFrames <= this.maxMissedFrames),
      ...fresh,
    ];
    return this.tracks
      .filter((track) => track.missedFrames === 0)
      .map((track) => ({ ...track, box: { ...track.box } }));
  }
  reset(): void {
    this.tracks = [];
    this.nextId = 1;
  }
}

File: apps/mobile/lib/features/scan/tracker.dart
import 'dart:math' as math;

import 'scan_models.dart';

double intersectionOverUnion(NormalizedBox a, NormalizedBox b) {
  final left = math.max(a.x, b.x);
  final top = math.max(a.y, b.y);
  final right = math.min(a.x + a.width, b.x + b.width);
  final bottom = math.min(a.y + a.height, b.y + b.height);
  final intersection =
      math.max(0.0, right - left) * math.max(0.0, bottom - top);
  final union = a.width * a.height + b.width * b.height - intersection;
  return union <= 0 ? 0 : intersection / union;
}

List<Detection> nonMaxSuppression(
  Iterable<Detection> detections, {
  double overlapThreshold = 0.45,
}) {
  final remaining = detections.toList()
    ..sort((a, b) => b.confidence.compareTo(a.confidence));
  final result = <Detection>[];
  while (remaining.isNotEmpty) {
    final candidate = remaining.removeAt(0);
    result.add(candidate);
    remaining.removeWhere(
      (other) =>
          other.label == candidate.label &&
          intersectionOverUnion(candidate.box, other.box) >= overlapThreshold,
    );
  }
  return result;
}

class CropTracker {
  final double matchIou;
  final int maxMissedFrames;
  List<CropTrack> _tracks = [];
  int _nextId = 1;

  CropTracker({this.matchIou = 0.15, this.maxMissedFrames = 4});

  List<CropTrack> update(Iterable<Detection> rawDetections) {
    final detections = nonMaxSuppression(rawDetections);
    final used = <int>{};
    final updated = <CropTrack>[];

    for (final track in _tracks) {
      var bestIndex = -1;
      var bestIou = matchIou;
      for (var index = 0; index < detections.length; index += 1) {
        if (used.contains(index)) continue;
        final iou = intersectionOverUnion(track.box, detections[index].box);
        if (iou > bestIou) {
          bestIou = iou;
          bestIndex = index;
        }
      }
      if (bestIndex < 0) {
        updated.add(track.missed());
      } else {
        used.add(bestIndex);
        final detection = detections[bestIndex];
        updated.add(
          CropTrack(
            id: track.id,
            label: detection.label,
            confidence: detection.confidence,
            box: detection.box,
          ),
        );
      }
    }

    for (var index = 0; index < detections.length; index += 1) {
      if (used.contains(index)) continue;
      final detection = detections[index];
      updated.add(
        CropTrack(
          id: _nextId++,
          label: detection.label,
          confidence: detection.confidence,
          box: detection.box,
        ),
      );
    }
    _tracks = updated
        .where((track) => track.missedFrames <= maxMissedFrames)
        .toList();
    return _tracks.where((track) => track.missedFrames == 0).toList();
  }

  void reset() {
    _tracks = [];
    _nextId = 1;
  }
}
